middleware/proxy/upstream.go

```diff
@@ -229,16 +232,32 @@ func parseBlock(c *caddyfile.Dispenser, u *staticUpstream) error {
 
 func (u *staticUpstream) healthCheck() {
 	for _, host := range u.Hosts {
-		port := ""
+		var hostName, checkPort string
+
+		// The DNS server might be an HTTP server.  If so, extract its name.
+		if url, err := url.Parse(host.Name); err == nil {
+			hostName = url.Host
+		} else {
+			hostName = host.Name
+		}
+
+		// Extract the port number from the parsed server name.
+		checkHostName, checkPort, err := net.SplitHostPort(hostName)
+		if err != nil {
+			checkHostName = hostName
+		}
+
 		if u.HealthCheck.Port != "" {
-			port = ":" + u.HealthCheck.Port
+			checkPort = u.HealthCheck.Port
 		}
-		hostURL := host.Name + port + u.HealthCheck.Path
+
+		hostURL := "http://" + net.JoinHostPort(checkHostName, checkPort) + u.HealthCheck.Path
 		if r, err := http.Get(hostURL); err == nil {
 			io.Copy(ioutil.Discard, r.Body)
 			r.Body.Close()
 			host.Unhealthy = r.StatusCode < 200 || r.StatusCode >= 400
 		} else {
+			fmt.Fprintf(os.Stderr, "Health check probe failed: %v\n", err)
```

Comment:
If an upstream ends up out of rotation due to a failing health check, having the check failures logged would be useful in performing root cause analysis. 

I couldn't find a more generic log infrastructure, other than the error logging middleware - and this is a warning, not an error.  Can you suggest a better way to deliver this diagnostic information to the user?